lib/slave/heartbeat.go

```diff
@@ -12,24 +12,43 @@ import (
 
 func (s *Slave) sendHeartbeat() {
 	log.Println("[Heartbeat] Sending...")
-	_, err := postHeartbeat(s)
+	heartbeat := s.generateHeartbeat()
+	_, err := s.postHeartbeat(heartbeat)
 
-	if err == nil {
-		var taskStatusses []lib.TaskStatus
-		for _, ts := range s.heartbeat.TaskStatus {
-			if ts.Status == lib.Running {
-				taskStatusses = append(taskStatusses, ts)
+	if err != nil {
+		log.Println("[Heartbeat] Delivery failed")
+	}
+}
+
+func (s *Slave) generateHeartbeat() lib.Heartbeat {
+	heartbeat := lib.Heartbeat{
+		SlaveId: s.id,
+	}
+	for i, task := range s.tasks {
+		var progress string
+		if task.Status == lib.Running {
+			c := make(chan string)
+			s.tasks[i].progressChan <- c
+			select {
+			case progress = <-c:
+			case <-time.After(time.Second * 1): // This could accually happen for a legitimate reason (much hash interations, slow hashing algorithms), but for now see what happens
+				log.Println("[ERROR]", "progressChan did not respond for task", task.ID)
 			}
 		}
-		s.heartbeat.TaskStatus = taskStatusses
-	} else {
-		log.Println("[Heartbeat] Delivery failed")
+		heartbeat.TaskStatus = append(heartbeat.TaskStatus, lib.TaskStatus{
+			Id:       task.ID,
+			JobId:    task.JobID,
+			Status:   task.Status,
+			Password: task.Password,
+			Progress: progress,
```

Comment:
If we fail to get the progress from the cracker then this would be the zero value. Can you put a comment here to remind ourselves that the master also needs to handle zero values in the progress?